sklearn/semi_supervised/tests/test_label_propagation.py

```diff
@@ -13,63 +14,70 @@
 from numpy.testing import assert_array_almost_equal
 from numpy.testing import assert_array_equal
 
-ESTIMATORS = [
-    (label_propagation.LabelPropagation, {'kernel': 'rbf'}),
-    (label_propagation.LabelPropagation, {'kernel': 'knn', 'n_neighbors': 2}),
-    (label_propagation.LabelPropagation, {
-        'kernel': lambda x, y: rbf_kernel(x, y, gamma=20)
-    }),
-    (label_propagation.LabelSpreading, {'kernel': 'rbf'}),
-    (label_propagation.LabelSpreading, {'kernel': 'knn', 'n_neighbors': 2}),
-    (label_propagation.LabelSpreading, {
+
+ESTIMATOR_CLS = [label_propagation.LabelPropagation,
+                 label_propagation.LabelSpreading]
+
+
+PARAMETERS = [{'kernel': 'rbf'}, {'kernel': 'knn', 'n_neighbors': 2}, {
         'kernel': lambda x, y: rbf_kernel(x, y, gamma=20)
-    }),
-]
+    }]
 
 
-def test_fit_transduction():
-    samples = [[1., 0.], [0., 2.], [1., 3.]]
+MATRIX_TYPES = (sp.csc_matrix, sp.csr_matrix, sp.coo_matrix, sp.dok_matrix,
+                sp.bsr_matrix, sp.lil_matrix, sp.dia_matrix, np.array)
+
+
+@pytest.mark.parametrize("matrix_type", MATRIX_TYPES)
+@pytest.mark.parametrize("estimator", ESTIMATOR_CLS)
+@pytest.mark.parametrize("parameters", PARAMETERS)
+def test_fit_transduction(matrix_type, estimator, parameters):
+    samples = matrix_type([[1., 0.], [0., 2.], [1., 3.]])
     labels = [0, 1, -1]
-    for estimator, parameters in ESTIMATORS:
-        clf = estimator(**parameters).fit(samples, labels)
-        assert clf.transduction_[2] == 1
+    clf = estimator(**parameters).fit(samples, labels)
+    assert clf.transduction_[2] == 1
 
 
-def test_distribution():
-    samples = [[1., 0.], [0., 1.], [1., 1.]]
+@pytest.mark.parametrize("matrix_type", MATRIX_TYPES)
+@pytest.mark.parametrize("estimator", ESTIMATOR_CLS)
+@pytest.mark.parametrize("parameters", PARAMETERS)
+def test_distribution(matrix_type, estimator, parameters):
+    samples = matrix_type([[1., 0.], [0., 1.], [1., 1.]])
     labels = [0, 1, -1]
-    for estimator, parameters in ESTIMATORS:
-        clf = estimator(**parameters).fit(samples, labels)
-        if parameters['kernel'] == 'knn':
-            continue    # unstable test; changes in k-NN ordering break it
-            assert_array_almost_equal(clf.predict_proba([[1., 0.0]]),
-                                      np.array([[1., 0.]]), 2)
-        else:
-            assert_array_almost_equal(np.asarray(clf.label_distributions_[2]),
-                                      np.array([.5, .5]), 2)
```

Comment:
We probably want to maintain both cases here for this test.

Reply:
Parameterisation removes the looping, so `continue` doesn't work. Plus, that test under the if seems unreachable to me even in the original function.

doing the following fails 16 tests.
```python
if parameters['kernel'] != 'knn':
        # unstable test; changes in k-NN ordering break it
        assert_array_almost_equal(clf.predict_proba([[1., 0.0]]),
                                  np.array([[1., 0.]]), 2)
        assert_array_almost_equal(np.asarray(clf.label_distributions_[2]),
                                  np.array([.5, .5]), 2)
```
Also, `dok` fails some convergence tests in `test_convergence_warning` and  `test_label_propagation_non_zero_normalizer`

I would appreciate some suggestions regarding how to proceed regarding the above issues. Thanks!

Reply:
Thank for the explanation, it makes sense.

Probably you might prefer using [`pytest.skip`](https://docs.pytest.org/en/reorganize-docs/new-docs/user/skipping.html) on such tests case.

> `dok` fails some convergence tests in `test_convergence_warning` and `test_label_propagation_non_zero_normalizer`

Do you have pointers to failing tests log?

Reply:
I am currently busy with others tasks and inspecting this bug will take a bit of a time as I am not familiar with this part of the code base yet. Thanks for reporting this and for your patience. 🙂

Reply:
I understand. No issues. 😄